internal/supervisor/installation.go

```diff
@@ -0,0 +1,340 @@
+	UnlockClusterInstallations(clusterInstallationID []string, lockerID string, force bool) (bool, error)
+	UpdateClusterInstallation(clusterInstallation *model.ClusterInstallation) error
+}
+
+// provisioner abstracts the provisioning operations required by the installation supervisor.
+type installationProvisioner interface {
+	CreateClusterInstallation(cluster *model.Cluster, installation *model.Installation, clusterInstallation *model.ClusterInstallation) error
+	DeleteClusterInstallation(cluster *model.Cluster, installation *model.Installation, clusterInstallation *model.ClusterInstallation) error
+}
+
+// InstallationSupervisor finds installations pending work and effects the required changes.
+//
+// The degree of parallelism is controlled by a weighted semaphore, intended to be shared with
+// other clients needing to coordinate background jobs.
+type InstallationSupervisor struct {
+	store       installationStore
+	provisioner installationProvisioner
+	instanceID  string
+	logger      log.FieldLogger
+}
+
+// NewInstallationSupervisor creates a new InstallationSupervisor.
+func NewInstallationSupervisor(store installationStore, installationProvisioner installationProvisioner, instanceID string, logger log.FieldLogger) *InstallationSupervisor {
+	return &InstallationSupervisor{
+		store:       store,
+		provisioner: installationProvisioner,
+		instanceID:  instanceID,
+		logger:      logger,
+	}
+}
+
+// Do looks for work to be done on any pending installations and attempts to schedule the required work.
+func (s *InstallationSupervisor) Do() error {
+	installations, err := s.store.GetUnlockedInstallationsPendingWork()
+	if err != nil {
+		s.logger.WithError(err).Warn("Failed to query for installation pending work")
+		return nil
+	}
+
+	for _, installation := range installations {
+		s.Supervise(installation)
+	}
+
+	return nil
+}
+
+// Supervise schedules the required work on the given installation.
+func (s *InstallationSupervisor) Supervise(installation *model.Installation) {
+	logger := s.logger.WithFields(map[string]interface{}{
+		"installation": installation.ID,
+	})
+
+	lock := newInstallationLock(installation.ID, s.instanceID, s.store, logger)
+	if !lock.TryLock() {
+		return
+	}
+	defer lock.Unlock()
+
+	logger.Debugf("Supervising installation in state %s", installation.State)
+
+	newState := s.transitionInstallation(installation, s.instanceID, logger)
+
+	installation, err := s.store.GetInstallation(installation.ID)
+	if err != nil {
+		logger.WithError(err).Warnf("failed to get installation and thus persist state %s", newState)
+		return
+	}
+
+	if installation.State == newState {
+		return
+	}
+
+	oldState := installation.State
+	installation.State = newState
+	err = s.store.UpdateInstallation(installation)
+	if err != nil {
+		logger.WithError(err).Warnf("Failed to set installation state to %s", newState)
+		return
+	}
+
+	logger.Debugf("Transitioned installation from %s to %s", oldState, newState)
```

Comment:
I wonder if this would be good to log at INFO level rather than DEBUG or would that be too spammy?

Reply:
I personally found this a little spammy, and tried to make sure that the INFO-level logs capture all necessary detail for still being able to understand the meaningful changes here. Propose we run debug enabled in production anyway, but filter out downstream.